presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -163,4 +260,59 @@ private static String writeValueAsString(Object value, ObjectMapper objectMapper
             throw new RuntimeException(e);
         }
     }
+
+    private <K, V> V get(LoadingCache<K, V> cache, K key)
+    {
+        try {
+            return cache.getUnchecked(key);
+        }
+        catch (UncheckedExecutionException e) {
+            throwIfInstanceOf(e.getCause(), PrestoException.class);
+            throw e;
+        }
+    }
+
+    private static class CacheKey<T>
+    {
+        private final SessionRepresentation session;
+        private final T key;
+
+        private CacheKey(SessionRepresentation session, T key)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.key = requireNonNull(key, "key is null");
+        }
+
+        public SessionRepresentation getSession()
+        {
+            return session;
+        }
+
+        public T getKey()
+        {
+            return key;
+        }
+
+        // Session changes across different queries. For caching to be effective across multiple queries,
+        // we should NOT include session in equals() and hashCode() methods below.
+        // Only the key value is to be considered in determining whether a cacheKey is unique or not
```

Comment:
Then why do we need to wrap such a big `SessionRepresentation` within a cache key? BTW, `SessionRepresentation` is going to be so big that leading to a lot of overhead of the cache.

Reply:
So the reason we have the sessionRepresentation into the cache key is because we need it to retrieve the result to cache upon a miss from the metadata provider. Basically first we check the cache for the result and if it misses it will go to the loading methods that will get the result to cache for us and they need the session value to do so even if it isn't part of checking the uniqueness of a key